### PR TITLE
Feature/cc 108

### PR DESCRIPTION
**Pull request by suNrisEinMyeYes**
- Beacon proxy for Cruzo1155
- tests for beacon
- scripts for beacon
- Factory for erc1155 tokens
- tests for factory
- scripts for factory

### CODE REVIEW

File: contracts/tokens/ERC1155CruzoBase.sol [hunk 1]
```diff
@@ -3,20 +3,21 @@
...
+    ContextUpgradeable,
+    ERC1155SupplyUpgradeable,
+    OwnableUpgradeable,
+    PausableUpgradeable
 {
     mapping(uint256 => address) public creators;
 
     /**
-     * @dev Require msg.sender to be the creator of the token id
+     * @dev Require masg.sender to be the creator of the token id
```
Comment (bobrom) [inline contracts/tokens/ERC1155CruzoBase.sol]: Unintentional modification

**Review by bobrom — CHANGES_REQUESTED** [review]
@suNrisEinMyeYes Please consider the following changes:
1. It's need to an emit event after new token has been created in Factory. Check with @kewka what data needs to be propagated over the event. 
2. Factory is not working for me, it throws the following error when I try to create a token: `Error: cannot estimate gas; transaction may fail or may require manual gas limit.`. I've created `deployToken` script to reproduce. 
3. I'd suggest to rename `Factory` contract into `Cruzo1155Factory` to make it more specific. Imagine, we want to have multiple factories in future. 
4. Please cover Factory contract with tests.

File: contracts/tokens/ERC1155CruzoBase.sol (hunk 1 above)
Comment (suNrisEinMyeYes) [reply contracts/tokens/ERC1155CruzoBase.sol]: fixed

File: contracts/utils/Cruzo1155Factory.sol [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/utils/Counters.sol";
+import "@openzeppelin/contracts/proxy/beacon/BeaconProxy.sol";
+import "@openzeppelin/contracts/utils/Context.sol";
+
+contract Cruzo1155Factory is Context, Ownable {
+    using Counters for Counters.Counter;
+    Counters.Counter private _tokenIds;
+
+    event newTokenCreated(
```
Comment (kewka) [inline contracts/utils/Cruzo1155Factory.sol]: https://docs.soliditylang.org/en/v0.8.15/style-guide.html#event-names

File: contracts/marketplace/CruzoMarket.sol [hunk 3]
```diff
@@ -101,64 +101,47 @@ contract CruzoMarket is
...
-    );
-    trades[_trade].status = "Executed";
-    ownerToTokenToItem[trade.poster][trade.tokenAddress][trade.itemId] = false;
-
-    trades[_trade].poster = payable(msg.sender);
-
-    emit TradeStatusChange(_trade, "Executed");
-  }
-
-  function buyItem(uint256 _trade, bytes calldata data) external payable {
```
Comment (kewka) [inline contracts/marketplace/CruzoMarket.sol:146]: Why has CruzoMarket been changed?

File: contracts/utils/Cruzo1155Factory.sol [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        address indexed tokenAddress,
+        uint256 indexed newTokenId
+    );
+
+    mapping(uint256 => address) private tokens;
```
Comment (kewka) [inline contracts/utils/Cruzo1155Factory.sol]: IMO, this is unnecessary

File: contracts/utils/Cruzo1155Factory.sol [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        string memory initBaseUri,
+        address initMarketAddress
+    ) {
+        beacon = _beacon;
+        selector = bytes4(keccak256(bytes(rawFuncInit)));
+        marketAddress = initMarketAddress;
+        baseUri = initBaseUri;
+    }
+
+    function create(string calldata _name, string calldata _symbol)
```
Comment (bobrom) [inline contracts/utils/Cruzo1155Factory.sol]: Please add `contractURI` argument for `create` function

File: contracts/utils/Cruzo1155Factory.sol (hunk 2 above)
Comment (suNrisEinMyeYes) [reply contracts/utils/Cruzo1155Factory.sol]: fixed, new name = NewTokenCreated

File: contracts/marketplace/CruzoMarket.sol (hunk 3 above)
Comment (suNrisEinMyeYes) [reply contracts/marketplace/CruzoMarket.sol:146]: Revert changes

File: contracts/utils/Cruzo1155Factory.sol (hunk 4 above)
Comment (suNrisEinMyeYes) [reply contracts/utils/Cruzo1155Factory.sol]: Have some troubles with testing now, i am used to get addr of token from this mapping, so, for short time i added variable called last, to track address to test.
In next commit i will rewrite this and events would be used

File: contracts/utils/Cruzo1155Factory.sol (hunk 5 above)
Comment (suNrisEinMyeYes) [reply contracts/utils/Cruzo1155Factory.sol]: added

**Review by bobrom — APPROVED** [review]

**Pull request merged**